hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala

```diff
@@ -1712,13 +1747,56 @@ object LowerTableIR {
         }
 
       case node =>
-        throw new LowererUnsupportedOperation(s"undefined: \n${ Pretty(ctx, node) }")
+        throw new LowererUnsupportedOperation(s"undefined: \n${Pretty(ctx, node)}")
     }
 
     assert(tir.typ.globalType == lowered.globalType, s"\n  ir global: ${tir.typ.globalType}\n  lowered global: ${lowered.globalType}")
     assert(tir.typ.rowType == lowered.rowType, s"\n  ir row: ${tir.typ.rowType}\n  lowered row: ${lowered.rowType}")
-    assert(lowered.key startsWith tir.typ.keyType.fieldNames, s"\n  ir key: ${tir.typ.keyType.fieldNames.toSeq}\n  lowered key: ${lowered.key}")
+    assert(tir.typ.keyType.isPrefixOf(lowered.kType), s"\n  ir key: ${tir.typ.key}\n  lowered key: ${lowered.key}")
 
     lowered
   }
+
+  /* We have a couple of options when repartitioning a table:
+   *  1. Send only the contexts needed to compute each new partition and
+   *     take/drop the rows that fall in that partition.
+   *  2. Compute the table with the old partitioner, write the table to cloud
+   *     storage then read the new partitions from the index.
+   *
+   * We'd like to do 1 as keeping things in memory (with perhaps a bit of work
+   * duplication) is generally less expensive than writing and reading a table
+   * to and from cloud storage. There comes a cross-over point, however, where
+   * it's cheaper to do the latter. One such example is as follows: consider a
+   * repartitioning where the same context is used to compute multiple
+   * partitions. The (parallel) computation of each partition involves at least
+   * all of the work to compute the previous partition:
+   *
+   *                  *----------------------*
+   *           in:    |                      |  ...
+   *                  *----------------------*
+   *                      /    |         \
+   *                     /     |          \
+   *                   *--*  *---*       *--*
+   *          out:     |  |  |   |  ...  |  |
+   *                   *--*  *---*       *--*
+   *
+   * We can estimate the relative cost of computing the new partitions vs
+   * spilling as being proportional to the mean number of old partitions
+   * used to compute new partitions.
+   */
+  def selectRepartitioning[A](original: RVDPartitioner, planned: RVDPartitioner)
```

Comment:
The inclusion of the recompute/reload args in the function seems to me like too much indirection. Can we just have this return a boolean and have the client branch itself?

An example of why I think this design is better is that we also want to gate the decision on an analysis that the consumer actually uses every partitions (no head/tail/filter_partitions) and there's actually no elegant way to do that with the function as-is. With a boolean, it's a simple `&&` in the client site.

Reply:
Happy to convert to a predicate, though you can just pass true and false to this function and achieve the same result.

Reply:
how so? You mean add it that analysis as a parameter here? Sure, but seems the most self-contained to just make this return the decision from partitioning alignment. Let's convert to the boolean predicate if you're ok with that.

Reply:
It's equivalent as we're just encoding boolean algebra with function:
```
selectRepartitioning(old, new)(ifTrue)(ifFalse)
if (selectReparitioning(old, new)(true)(false)) ifTrue else ifFalse
if (selectReparitioning(old, new)(true)(false) && condition) ifTrue else ifFalse
if (selectReparitioning(old, new)(condition)(false)) ifTrue else ifFalse
```

Reply:
agree it's semantically equivalent, I just find this one clearest:
```
if (selectReparitioning(old, new)(true)(false) && condition) ifTrue else ifFalse
```